spring-cloud-dataflow-core/src/main/java/org/springframework/cloud/dataflow/core/dsl/GraphGeneratorVisitor.java

```diff
@@ -0,0 +1,651 @@
+			GraphGeneratorVisitor.Sequence sequence, List<Context.TransitionTarget> transitionTargets) {
+		
+		// Record a map of ids in the sequence to the new node ids to use when creating new links
+		Map<String, String> nodeIds = new HashMap<>();
+
+		// Create copies of all the nodes - except the START and END
+		Node startNode = sequence.nodes.get(0);
+		Node endNode = sequence.nodes.get(sequence.nodes.size() - 1);
+		for (int i = 1; i < (sequence.nodes.size() - 1); i++) {
+			Node n = sequence.nodes.get(i);
+			Node newNode = new Node(nextId(), n.name, n.properties);
+			nodeIds.put(n.id, newNode.id);
+			mainSequence.nodes.add(newNode);
+		}
+		
+		// Now copy links
+		for (int i = 0; i < sequence.links.size(); i++) {
+			Link existingLink = sequence.links.get(i);
+			String existingLinkFrom = existingLink.from;
+			String existingLinkTo = existingLink.to;
+			Link newLink = null;
+			if (existingLinkFrom.equals(startNode.id)) {
+				// This link need replacing with links from the transition points to the same
+				// target
+				for (Context.TransitionTarget tt : transitionTargets) {
+					newLink = new Link(tt.nodeId, nodeIds.get(existingLinkTo), tt.onState);
+					mainSequence.links.add(newLink);
+				}
+			} else if (existingLinkTo.equals(endNode.id)) {
+				// This link needs replacing with links from the final copied node in the sequence
+				// to the end node in the main sequence
+				Context.TransitionTarget tt = transitionTargets.get(0);
+				// assert all the transitionTargets have the same lastNodeId
+				String finalNodeInMainSequence = tt.lastNodeId;
+				List<Link> newLinks = new ArrayList<>();
+				for (Link l : mainSequence.links) {
+					if (l.from.equals(finalNodeInMainSequence)) {
+						newLinks.add(new Link(nodeIds.get(existingLinkFrom), l.to, l.getTransitionName()));
+					}
+				}
+				mainSequence.links.addAll(newLinks);
+			} else {
+				newLink = new Link(nodeIds.get(existingLinkFrom), nodeIds.get(existingLinkTo),
+						existingLink.getTransitionName());
+				mainSequence.links.add(newLink);
+			}
+		}
+		// After inlining the sequence, the mainSequence may have inherited new
+		// outstanding transitions
+		if (DEBUG) {
+			System.out.println("Main sequence inheriting outstanding transitions: "+sequence.outstandingTransitions);
+		}
+		List<Context.TransitionTarget> rewrittenTransitions = new ArrayList<>();
+		for (Context.TransitionTarget looseEnd : sequence.outstandingTransitions) {
+			Context.TransitionTarget tt = new Context.TransitionTarget(nodeIds.get(looseEnd.nodeId), looseEnd.onState,
+					looseEnd.label);
+			// They should have the same 'flow' as the sequence they are injected into
+			tt.flow = transitionTargets.get(0).flow;
+			tt.lastNodeId = transitionTargets.get(0).lastNodeId;//nodeIds.get(looseEnd.lastNodeId);
+			if (DEBUG) {
+				System.out.println("Remapped "+looseEnd+" to "+tt);
+			}
+			rewrittenTransitions.add(tt);
+		}
+		mainSequence.outstandingTransitions.addAll(rewrittenTransitions);
```

Comment:
Why not add tt's directly to mainSequence.outstandingTransitions instead of adding them to the rewrittenTransitions list and doing a addAll.